### PR TITLE
Handle collector view

### LINKED ISSUE #4749: Collectors subform opens as read only and you can't change it
**Describe the bug**
The collectors subform opens as read only and expanded

**To Reproduce**

1. Go to collecting event/information 
2. Add a collector
3. See problem 

**Expected behavior**
You should be able to edit it and it should not open as read only. 

**Screenshots**

https://github.com/specify/specify7/assets/150940840/848a611e-4b27-414f-8813-868f191a4063

Please, also fill out the following information manually:

- OS: Windows 11
- Browser: Chrome
- Specify 7 Version: edge
- Database Name: lsumzmammals9823
- Collection name: mammals
- User Name: SpAdmin
[Specify 7 System Information - 2024-04-05T18_16_21.848Z.txt](https://github.com/specify/specify7/files/14889293/Specify.7.System.Information.-.2024-04-05T18_16_21.848Z.txt)

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #4749

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add relevant issue to release milestone

### Testing instructions
Go to collecting event/information
Add a collector

- [ ] See valid form displays

### CODE REVIEW

File: specifyweb/frontend/js_src/lib/components/FormCells/FormTable.tsx [hunk 1]
```diff
@@ -169,6 +171,16 @@ export function FormTable<SCHEMA extends AnySchema>({
     mode,
   });
 
+  const validExpandedViewDefinition = viewToChange.has(viewName)
+    ? useViewDefinition({
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/FormCells/FormTable.tsx:175]: you can't call hooks conditionally
how many times??? 😁

**Review by maxpatiiuk — CHANGES_REQUESTED** [review]

File: specifyweb/frontend/js_src/lib/components/FormCells/FormTable.tsx (hunk 1 above)
Comment (CarolineDenis) [reply specifyweb/frontend/js_src/lib/components/FormCells/FormTable.tsx:175]: :grimacing: so sorry

**Comment by CarolineDenis** [comment]
todo: 
change name of variables

**Review by maxpatiiuk — COMMENTED** [review]
could you please investigate why the Collectors view in particular is exhibiting this issue?
it might be that Collectors is not the only place where we have this issue.
is this related to the fix for https://github.com/specify/specify7/issues/4737?

we are already fetching 2 form definitions here. fetching a 3rd is one too many 😊

**Comment by melton-jason** [comment]
> could you please investigate why the Collectors view in particular is exhibiting this issue? it might be that Collectors is not the only place where we have this issue. is this related to the fix for #4737?
> 
> we are already fetching 2 form definitions here. fetching a 3rd is one too many 😊


@maxpatiiuk 

There is a breakdown going over why this issue happens with Collectors in https://github.com/specify/specify7/issues/4749#issuecomment-2040440456. 

TLDR: 
The default CollectingEvent subview specifies the `Collectors` view should be used. 
Within the `Collectors` view, there are the fields `agent.firstname` and `agent.lastname` which appear on the form as readonly. 
This issue had not occurred earlier because before https://github.com/specify/specify7/pull/4740, Specify was using the `table.view` of the table to define which expanded form definition to use (in this case `Collector`).
After #4740, Specify uses the view specified in the subview for both the collapsed and expanded views.

**Comment by melton-jason** [comment]
There may be a generic solution for other possible forms which may be affected (already, the GroupPerson subview in Agent has been shown to exhibit the same behavior as the Collector views: https://github.com/specify/specify7/issues/4749#issuecomment-2045915336)

The ideal fix would be to modify the default views, but since that can not occur until we separate from Specify 6, thankfully there seems to be enough consistency in cases for which this issue can be resolved. 

Here are some examples of default views and their view definitions:

---

Default Preparation views for Fish Managers: 

[Fish Manager Preparation View](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/fish/manager/fish.views.xml#L112-L123)
[Fish Manager Preparation View Defs](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/fish/manager/fish.views.xml#L618-L715)

The returned json from `/context/view.json?name=Preparation`:
```json
{
   "name":"Preparation",
   "class":"edu.ku.brc.specify.datamodel.Preparation",
   "busrules":"edu.ku.brc.specify.datamodel.busrules.PreparationBusRules",
   "resourcelabels":"false",
   "altviews":{
      "Preparation Table View":{
         "name":"Preparation Table View",
         "viewdef":"Preparation Table",
         "mode":"view"
      },
      "Preparation Table Edit":{
         "name":"Preparation Table Edit",
         "viewdef":"Preparation Table",
         "mode":"edit"
      },
      "Preparation View":{
         "name":"Preparation View",
         "viewdef":"Preparation",
         "mode":"view"
      },
      "Preparation Edit":{
         "name":"Preparation Edit",
         "viewdef":"Preparation",
         "mode":"edit",
         "default":"true"
      }
   },
   "viewdefs":{
      "Preparation":"<viewdef type=\"form\" name=\"Preparation\" class=\"edu.ku.brc.specify.datamodel.Preparation\" gettable=\"edu.ku.brc.af.ui.forms.DataGetterForObj\" settable=\"edu.ku.brc.af.ui.forms.DataSetterForObj\" useresourcelabels=\"true\">\n            <desc>Preparation for Collection Object form.</desc>\n            <enableRules />\n\n            <columnDef>100px,2px,195px,5px,91px,2px,60px,5px,83px,2px,60px,5px,85px,2px,123px,2px,70px,p:g</columnDef>\n            <columnDef os=\"lnx\">115px,2px,200px,5px,110px,2px,67px,5px,82px,2px,67px,5px,100px,2px,126px,2px,75px,p:g</columnDef>\n            <columnDef os=\"mac\">130px,2px,215px,5px,110px,2px,80px,5px,93px,2px,80px,5px,105px,2px,174px,2px,70px,p:g</columnDef>\n            <columnDef os=\"exp\">p,2px,p:g(2),5px:g,p,2px,p:g,5px:g,p,2px,p:g,5px:g,p,2px,p:g,10px,p:g</columnDef>\n            <rowDef auto=\"true\" cell=\"p\" sep=\"2dlu\" />\n\n            <rows>\n                <row>\n                    <cell type=\"label\" labelfor=\"7\" />\n                    <cell type=\"field\" id=\"7\" name=\"prepType\" uitype=\"combobox\" picklist=\"PrepType\" />\n                    <cell type=\"label\" labelfor=\"3\" />\n                    <cell type=\"field\" id=\"3\" name=\"countAmt\" uitype=\"text\" />\n\t\t\t\t\t<cell type=\"label\" label=\"Current Count:\" initialize=\"align=right\" />\n\t\t\t\t\t<cell type=\"field\" id=\"ti\" name=\"actualCountAmt\" uitype=\"text\" readonly=\"true\" />\n                    <cell type=\"field\" id=\"isOnLoan\" name=\"isOnLoan\" uitype=\"checkbox\" label=\"IS_ON_LOAN\" readonly=\"true\" />\n                    <cell type=\"command\" id=\"ShowLoansBtn\" name=\"ShowLoansBtn\" label=\"SHOW_INTERACTIONS\" ignore=\"true\" />\n                    <cell type=\"subview\" id=\"attachments\" viewname=\"ObjectAttachment\" name=\"preparationAttachments\" initialize=\"align=right;btn=true;icon=PreparationAttachment\" />\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"2\" />\n                    <cell type=\"field\" id=\"2\" name=\"preparedByAgent\" uitype=\"querycbx\" initialize=\"name=Agent\" />\n                    <cell type=\"label\" labelfor=\"1\" />\n                    <cell type=\"field\" id=\"1\" uitype=\"plugin\" name=\"this\" initialize=\"name=PartialDateUI;df=preparedDate;tp=preparedDatePrecision\" uifieldformatter=\"Date\" colspan=\"5\" />\n                    <cell type=\"label\" labelfor=\"4\" />\n                    <cell type=\"field\" id=\"4\" name=\"storage\" uitype=\"querycbx\" initialize=\"name=Storage\" colspan=\"3\" />\n                </row>\n                <row>\n                    <cell type=\"subview\" viewname=\"DeaccessionPreparation2\" id=\"deaccprep\" name=\"deaccessionPreparations\" colspan=\"12\" />\n                </row>\n                <!--<row>\n                    <cell type=\"label\" labelfor=\"desc\"/>\n                    <cell type=\"field\" id=\"desc\" name=\"description\" uitype=\"text\"/>\n                    <cell type=\"label\" labelfor=\"n1\"/>\n                    <cell type=\"field\" id=\"n1\" name=\"number1\" uitype=\"text\"/>\n                    <cell type=\"label\" labelfor=\"n2\"/>\n                    <cell type=\"field\" id=\"n2\" name=\"number2\" uitype=\"text\"/>\n                </row>\n                <row>\n                    <cell type=\"subview\" id=\"ms\" viewname=\"MaterialSample\" name=\"materialSamples\" initialize=\"align=right;btn=true;icon=MaterialSample\"/>\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"rem\"/>\n                    <cell type=\"field\" id=\"rem\" name=\"remarks\" uitype=\"textareabrief\" rows=\"2\" colspan=\"14\"/>\n                </row> \n                <row>\n                    <cell type=\"label\" labelfor=\"stat\"/>\n                    <cell type=\"field\" id=\"stat\" name=\"status\" uitype=\"text\"/>\n                    <cell type=\"label\" labelfor=\"sl\"/>\n                    <cell type=\"field\" id=\"sl\" name=\"storageLocation\" uitype=\"text\"/>\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"t1\"/>\n                    <cell type=\"field\" id=\"t1\" name=\"text1\" uitype=\"text\"/>\n                    <cell type=\"label\" labelfor=\"t2\"/>\n                    <cell type=\"field\" id=\"t2\" name=\"text2\" uitype=\"text\"/>\n                </row>\n                <row>\n                    <cell type=\"field\" id=\"yn1\" name=\"yesNo1\" uitype=\"checkbox\" initialize=\"editable=true\"/>\n                    <cell type=\"field\" id=\"yn2\" name=\"yesNo2\" uitype=\"checkbox\" initialize=\"editable=true\"/>\n                    <cell type=\"field\" id=\"yn3\" name=\"yesNo3\" uitype=\"checkbox\" initialize=\"editable=true\"/>\n                </row>\n                <row>\n                    <cell type=\"subview\" id=\"18\" viewname=\"PreparationAttributes\" name=\"preparationAttribute\" colspan=\"15\"/>\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"9\"/>\n                    <cell type=\"field\" id=\"9\" name=\"createdByAgent\" uitype=\"label\" readonly=\"true\" uifieldformatter=\"Agent\"/>\n                    <cell type=\"label\" labelfor=\"10\"/>\n                    <cell type=\"field\" id=\"10\" name=\"modifiedByAgent\" uitype=\"label\" readonly=\"true\" uifieldformatter=\"Agent\"/>\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"11\"/>\n                    <cell type=\"field\" id=\"11\" name=\"timestampModified\" uitype=\"label\" readonly=\"true\"/>\n                    <cell type=\"label\" labelfor=\"12\"/>\n                    <cell type=\"field\" id=\"12\" name=\"timestampCreated\" uitype=\"label\" readonly=\"true\"/>\n                </row>-->\n            </rows>\n        </viewdef>\n        \n        ",
      "Preparation Table":"<viewdef type=\"formtable\" name=\"Preparation Table\" class=\"edu.ku.brc.specify.datamodel.Preparation\" gettable=\"edu.ku.brc.af.ui.forms.DataGetterForObj\" settable=\"edu.ku.brc.af.ui.forms.DataSetterForObj\">\n            <desc>Preparation subform table for Collection Object form.</desc>\n            <definition>Preparation</definition>\n        </viewdef>\n                \n        "
   },
   "view":"<view name=\"Preparation\" class=\"edu.ku.brc.specify.datamodel.Preparation\" busrules=\"edu.ku.brc.specify.datamodel.busrules.PreparationBusRules\" resourcelabels=\"false\">\n            <desc>Subform within the Preparation form.</desc>\n            <altviews>\n                <altview name=\"Preparation Table View\" viewdef=\"Preparation Table\" mode=\"view\" />\n                <altview name=\"Preparation Table Edit\" viewdef=\"Preparation Table\" mode=\"edit\" />\n                <altview name=\"Preparation View\" viewdef=\"Preparation\" mode=\"view\" />\n                <altview name=\"Preparation Edit\" viewdef=\"Preparation\" mode=\"edit\" default=\"true\" />\n            </altviews>\n        </view>\n        \n    ",
   "viewsetName":"Fish Views",
   "viewsetLevel":"UserType",
   "viewsetSource":"disk",
   "viewsetId":null,
   "viewsetFile":"fish/manager/fish.views.xml"
}
```
[Collector Views](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/common/common.views.xml#L205-L214)
[Collector View Defs](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/common/common.views.xml#L2400-L2442)

The returned json from `/context/view.json?name=Collector`:
```json
{
   "name":"Collector",
   "class":"edu.ku.brc.specify.datamodel.Collector",
   "busrules":"edu.ku.brc.specify.datamodel.busrules.CollectorBusRules",
   "altviews":{
      "Collector View":{
         "name":"Collector View",
         "viewdef":"Collector",
         "mode":"view",
         "validated":"false",
         "default":"true"
      },
      "Collector Edit":{
         "name":"Collector Edit",
         "viewdef":"Collector",
         "mode":"edit",
         "validated":"true"
      }
   },
   "viewdefs":{
      "Collector":"<viewdef type=\"form\" name=\"Collector\" class=\"edu.ku.brc.specify.datamodel.Collector\" gettable=\"edu.ku.brc.af.ui.forms.DataGetterForObj\" settable=\"edu.ku.brc.af.ui.forms.DataSetterForObj\">\n            <desc>The Collector form.</desc>\n            <enableRules />\n\n            <columnDef>105px,2px,210px,5px,100px,2px,98px,300px,p:g</columnDef>\n            <columnDef os=\"lnx\">135px,2px,230px,5px,120px,2px,118px,325px,p:g</columnDef>\n            <columnDef os=\"mac\">130px,2px,215px,5px,140px,2px,138px,395px,p:g</columnDef>\n            <columnDef os=\"exp\">p,2px,p:g,5px:g,p,2px,p:g,p:g(2),p:g</columnDef>\n            <rowDef>p,2dlu,p:g,2dlu,p:g</rowDef>\n\n            <rows>\n                <row>\n                    <cell type=\"label\" labelfor=\"1\" />\n                    <cell type=\"field\" id=\"1\" name=\"agent\" uitype=\"querycbx\" initialize=\"name=Agent;title=Agent\" />\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"3\" />\n                    <cell type=\"field\" id=\"3\" name=\"remarks\" uitype=\"textareabrief\" rows=\"2\" colspan=\"6\" />\n                </row>\n                <!--<row>\n                    <cell type=\"label\" labelfor=\"9\"/>\n                    <cell type=\"field\" id=\"9\" name=\"createdByAgent\" uitype=\"label\" readonly=\"true\"  uifieldformatter=\"Agent\"/>\n                    <cell type=\"label\" labelfor=\"10\"/>\n                    <cell type=\"field\" id=\"10\" name=\"modifiedByAgent\" uitype=\"label\" readonly=\"true\"  uifieldformatter=\"Agent\"/>\n                    <cell type=\"label\" id=\"divLabel\" label=\" \" initialize=\"align=right\"/>\n                    <cell type=\"field\" id=\"4\" name=\"divisionCBX\" uitype=\"combobox\" ignore=\"true\"/>\n                </row>\n                <row>\n                \t<cell type=\"field\" id=\"34\" uitype=\"checkbox\" name=\"isPrimary\"/>                \t\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"11\"/>\n                    <cell type=\"field\" id=\"11\" name=\"timestampModified\" uitype=\"label\" readonly=\"true\"/>\n                    <cell type=\"label\" labelfor=\"12\"/>\n                    <cell type=\"field\" id=\"12\" name=\"timestampCreated\" uitype=\"label\" readonly=\"true\"/>\n                </row>-->\n            </rows>\n        </viewdef>\n\n        "
   },
   "view":"<view name=\"Collector\" class=\"edu.ku.brc.specify.datamodel.Collector\" busrules=\"edu.ku.brc.specify.datamodel.busrules.CollectorBusRules\">\n            <desc>The Collectors Form.</desc>\n            <altviews>\n                <altview name=\"Collector View\" viewdef=\"Collector\" mode=\"view\" validated=\"false\" default=\"true\" />\n                <altview name=\"Collector Edit\" viewdef=\"Collector\" mode=\"edit\" validated=\"true\" />\n            </altviews>\n        </view>\n\n        ",
   "viewsetName":"Common",
   "viewsetLevel":"Common",
   "viewsetSource":"disk",
   "viewsetId":null,
   "viewsetFile":"common/common.views.xml"
}
```

[Collectors View](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/common/common.views.xml#L216-L227)
[Collectors View Defs](https://github.com/specify/specify6/blob/945d28f00288d2d64979848fb2fb63fda34bdd1b/config/common/common.views.xml#L2444-L2489)

The returned json from `/context/view.json?name=Collectors`:
```json
{
   "name":"Collectors",
   "class":"edu.ku.brc.specify.datamodel.Collector",
   "busrules":"edu.ku.brc.specify.datamodel.busrules.CollectorBusRules",
   "altviews":{
      "Collectors Table View":{
         "name":"Collectors Table View",
         "viewdef":"Collectors Table",
         "mode":"view"
      },
      "Collectors Table Edit":{
         "name":"Collectors Table Edit",
         "viewdef":"Collectors Table",
         "mode":"edit",
         "default":"true"
      }
   },
   "viewdefs":{
      "Collectors Table":"<viewdef type=\"formtable\" name=\"Collectors Table\" class=\"edu.ku.brc.specify.datamodel.Collector\" gettable=\"edu.ku.brc.af.ui.forms.DataGetterForObj\" settable=\"edu.ku.brc.af.ui.forms.DataSetterForObj\">\n            <desc>Collectors grid view for CollectingEvent form.</desc>\n            <definition>Collectors</definition>\n        </viewdef>\n\n        ",
      "Collectors":"<viewdef type=\"form\" name=\"Collectors\" class=\"edu.ku.brc.specify.datamodel.Collector\" gettable=\"edu.ku.brc.af.ui.forms.DataGetterForObj\" settable=\"edu.ku.brc.af.ui.forms.DataSetterForObj\">\n            <desc>The Collectors form - UNKNOWN use in database.</desc>\n            <enableRules />\n\n            <columnDef>p,5dlu,p:g,5dlu,p</columnDef>\n            <rowDef auto=\"true\" cell=\"p\" sep=\"2px\" />\n\n            <rows>\n                <row>\n                    <cell type=\"label\" labelfor=\"3\" />\n                    <cell type=\"field\" id=\"3\" name=\"agent.lastName\" uitype=\"text\" colspan=\"3\" />\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"5\" />\n                    <cell type=\"field\" id=\"5\" name=\"agent.firstName\" uitype=\"text\" colspan=\"3\" />\n                </row>\n                <row>\n                    <cell type=\"label\" labelfor=\"7\" />\n                    <cell type=\"field\" id=\"7\" name=\"remarks\" uitype=\"text\" colspan=\"3\" />\n                </row>\n            </rows>\n        </viewdef>\n\n        "
   },
   "view":"<view name=\"Collectors\" class=\"edu.ku.brc.specify.datamodel.Collector\" busrules=\"edu.ku.brc.specify.datamodel.busrules.CollectorBusRules\">\n            <desc>The Collectors Subform.</desc>\n            <altviews>\n                <!-- <altview name=\"Collectors Icon View\"  viewdef=\"CollectorsIconView\" mode=\"view\"/>\n                <altview name=\"Collectors Icon Edit\"  viewdef=\"CollectorsIconView\" mode=\"edit\"/>  -->\n                <altview name=\"Collectors Table View\" viewdef=\"Collectors Table\" mode=\"view\" />\n                <altview name=\"Collectors Table Edit\" viewdef=\"Collectors Table\" mode=\"edit\" default=\"true\" />\n            </altviews>\n        </view>\n\n        ",
   "viewsetName":"Common",
   "viewsetLevel":"Common",
   "viewsetSource":"disk",
   "viewsetId":null,
   "viewsetFile":"common/common.views.xml"
}
```
(The `GroupPersons` and `GroupPerson` views follow the same structure as Collectors/Collector) 

---

In the default forms, it seems there are two separate ways to represent a FormTable. 
The first and expected way is the case like `Preparation`, in which the Grid and Form view definitions are defined in a single view. 
The other way which is present is to have two separate views, where each view and their view definitions separately represent the Grid and Form views in the form. 
This causes a problem in Specify 7 which expects all view definitions relating to a view to be associated with the view and not split amongst multiple views.


### Solution
(Unless there is a third way of representing FormTables...)

Thankfully there is some consistency amongst the table-only views. 
Notably, every `altview` in the `altviews` of the view points to a view definition which is of type `formtable`.
When reading a form definition for a subview on the frontend, if every altview points to a `formtable` viewdef in a view, we can likely fallback to using the `table.view` view for rendering the expanded definition of the subview (the collapsed view can still remain the original view).

**Comment by CarolineDenis** [comment]
After meeting with @melton-jason, 
We decided to go with the array solution for now. We will change this behavior once we are able to modify the default views. 
Steps: 
- go through the 6 default files and identify the table that are concerned by this issue 
- add them to the array [viewToChange]

**Comment by melton-jason** [comment]
> After meeting with @melton-jason, We decided to go with the array solution for now. We will change this behavior once we are able to modify the default views. Steps:
> 
> * go through the 6 default files and identify the table that are concerned by this issue
> * add them to the array [viewToChange]

One caveat with this approach is that we would be overriding any user-defined views if they had names which were included in the array. 
(Essentially artificially inducing https://github.com/specify/specify7/issues/4737). 

It may be worth considering altering the `useViewDefinition  hook (and other functions which the hook uses) and the `ViewDescription` type to include or consider altviews for this case.

**Comment by maxpatiiuk** [comment]
@melton-jason @CarolineDenis based on a git blame, here is a commit that modified the relevant code
https://github.com/specify/specify7/commit/d7f40a72225b87124283d636d8bc938e1fb80a0e

The commit also mentions:
> The fix for https://github.com/specify/specify7/issues/2366 caused https://github.com/specify/specify7/issues/2412
This commit reverts that fix and adds a better fix

Those issues contain additional context that might help you in addressing this issue better.
We could also check how this is supposed to work in sp6 - maybe there is something we are missing here

---

> add them to the array [viewToChange]

Jason described a heuristic to help decide what views should be included in that "viewToChange". Could that be used to avoid hardcoding the list of view names?

> Notably, every altview in the altviews of the view points to a view definition which is of type formtable.
When reading a form definition for a subview on the frontend, if every altview points to a formtable viewdef in a view, we can likely fallback to using the table.view view for rendering the expanded definition of the subview (the collapsed view can still remain the original view).

**Comment by CarolineDenis** [comment]
@maxpatiiuk, 

> Jason described a heuristic to help decide what views should be included in that "viewToChange". Could that be used to avoid hardcoding the list of view names?

yes we could have a way of doing that but we would need to change the useViewDefinition hook (and other functions which the hook uses) and the ViewDescription type to include or consider altviews for this case

**Comment by melton-jason** [comment]
I have opened an alternative PR (#4776) with the proposed change from https://github.com/specify/specify7/pull/4760#issuecomment-2046022176

**Comment by CarolineDenis** [comment]
fixed in PR (https://github.com/specify/specify7/pull/4776)
